src/vsr/routing.zig

```diff
@@ -0,0 +1,877 @@
+    // Maximum latency corresponds to the time when a prepare is fully replicated, and is important
+    // for the overall health of the cluster.
+    maximum: Duration,
+    // Worst-case latency tracks the length of the critical path, but we want non-critical paths to
+    // be as short as possible.
+    sum: Duration,
+
+    fn less(lhs: Cost, rhs: Cost) bool {
+        // For sum, smaller than 5% improvement is considered insignificant.
+        // For median and maximum the threshold is 10%.
+        //
+        // Why 5%: If replicas are on a ring in terms of distance, the sum for optimal path is
+        //
+        //   1 → 2
+        //  ↗
+        // ♔       3
+        //  ↘     ↗
+        //   5 → 4
+        //
+        // sum = (1 + 1) + (1 + 1 + 1)   (prepares)
+        //       + (1 + 2) + (1 + 2 + 3) (prepare ok)
+        //     = 14
+        //
+        // The sum for next best path is
+        //
+        //   1 → 2
+        //  ↗
+        // ♔       3
+        //  ↘  ↗  ↙
+        //   5   4
+        //
+        // sum = (1 + 1) + (1 + 2 + 1)   (prepares, two replicas transposed)
+        //       + (1 + 2) + (1 + 2 + 3) (prepare ok)
+        //     = 15
+        //
+        // The difference between 14 and 15 is less than 10% but more than 5%.
+        //
+        // Why 10%: just feels like a reasonable number! median varies more than sum, so we need
+        // larger tolerance there.
+        inline for (.{
+            .{ "median", ratio(1, 10) },
+            .{ "maximum", ratio(1, 10) },
+            .{ "sum", ratio(1, 20) },
+        }) |field_threshold| {
+            const field, const threshold = field_threshold;
+            if (less_signifiantly(@field(lhs, field), @field(rhs, field), threshold)) return true;
+            if (@field(lhs, field).ns > @field(rhs, field).ns) return false;
+        }
+        return false;
+    }
+
+    // Returns true if lhs + lhs⋅threshold < rhs.
+    fn less_signifiantly(lhs: Duration, rhs: Duration, thershold: Ratio) bool {
+        assert(thershold.numerator < thershold.denominator);
+        return lhs.ns * (thershold.numerator + thershold.denominator) <
+            rhs.ns * thershold.denominator;
+    }
+
+    fn avg(lhs: Cost, rhs: Cost) Cost {
+        return .{
+            .median = .{ .ns = @divFloor(lhs.median.ns + rhs.median.ns, 2) },
+            .maximum = .{ .ns = @divFloor(lhs.maximum.ns + rhs.maximum.ns, 2) },
+            .sum = .{ .ns = @divFloor(lhs.sum.ns + rhs.sum.ns, 2) },
+        };
+    }
+
+    fn ewma_add(old: Cost, new: Cost) Cost {
+        return .{
+            .median = ewma_add_duration(old.median, new.median),
+            .maximum = ewma_add_duration(old.maximum, new.maximum),
+            .sum = ewma_add_duration(old.sum, new.sum),
+        };
+    }
+
+    fn ewma_add_duration(old: Duration, new: Duration) Duration {
+        return .{
+            .ns = @divFloor((old.ns * 4) + new.ns, 5),
+        };
+    }
+};
+
+const Routing = @This();
```

Comment:
I am trying out a new pattern where 

```
count This = @This();
```


Is the last type defined, which separates fields&types from methods.